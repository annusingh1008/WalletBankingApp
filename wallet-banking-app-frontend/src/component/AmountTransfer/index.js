import React, { useState } from "react";
import Layout from "../Layout";
import { Row, Col, Container, Button, Form } from "react-bootstrap";
import { useDispatch, useSelector } from "react-redux";
import { amountTransfer } from "../../actions/amountTransfer.action";
import { getBalance } from "../../reducers/balance.reducer";

const AmountTransfer = () => {
  const [amount, setAmount] = useState();
  const [email, setEmail] = useState("");
  const [amountErr, setAmountErr] = useState("");
  const [isAmountSufficient, setIsAmountSufficient] = useState("");

  const dispatch = useDispatch();
  const balance = useSelector(getBalance);
  const userEmail = localStorage.getItem("email");

  const confirmAmountTransfer = (e) => {
    e.preventDefault();
    if (amount <= 0) {
      setIsAmountSufficient("");
      setAmountErr("Please enter a valid amount");
    } else if (amount > balance.amount) {
      setAmountErr("");
      setIsAmountSufficient("Insufficient Amount");
    } else {
      const amountDetails = {
        email: userEmail,
        amount: amount,
        creditToEmail: email,
        currentAmount: balance.amount,
      };
      dispatch(amountTransfer(amountDetails));
    }
  };

  return (
    <Layout sidebar>
      <Container className="container">
        <Row md={12}>
          <Col>
            <div className="heading">Transfer Amount</div>
            <Form
              data-testid="form"
              className="form"
              onSubmit={(e) => confirmAmountTransfer(e)}
            >
              <label>Email</label>
              <br />

              <input
                data-testid="emailInputBox"
                type="email"
                placeholder="Email"
                value={email}
                onChange={(e) => setEmail(e.target.value)}
                required
              ></input>
              <br />

              <label>Amount</label>
              <br />

              <input
                data-testid="amountInputBox"
                type="text"
                placeholder="1000"
                value={amount}
                onChange={(e) => setAmount(e.target.value)}
                required
              ></input>
              <br />
              {amountErr && (
                <p data-testid="error-msg" className="text-danger">
                  {amountErr}
                </p>
              )}
              {isAmountSufficient && (
                <p data-testid="insufficient-amount" className="text-danger">
                  Insufficient Amount
                </p>
              )}

              <Button data-testid="button" type="variant">
                Transfer Amount
              </Button>
            </Form>
          </Col>
        </Row>
      </Container>
    </Layout>
  );
};

export default AmountTransfer;
