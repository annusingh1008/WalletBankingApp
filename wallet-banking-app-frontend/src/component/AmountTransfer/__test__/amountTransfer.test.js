import React from "react";
import { render, screen, fireEvent, act } from "@testing-library/react";
import AmountTransfer from "../index";
import { Provider } from "react-redux";
import store from "../../../store/index";
import { BrowserRouter as Router } from "react-router-dom";
import { getBalance } from "../../../reducers/balance.reducer";
import { Container, Form } from "react-bootstrap";
import { amountTransfer } from "../../../actions/amountTransfer.action";

const mockDispatch = jest.fn();
jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useDispatch: () => mockDispatch,
}));

jest.mock("../../../actions/amountTransfer.action", () => ({
  amountTransfer: jest.fn(() => ({ type: "amountTransfer" })),
}));

jest.mock("../../../reducers/balance.reducer", () => ({
  getBalance: jest.fn(() => ({ amount: 1000 })),
}));

jest.mock("../../Layout", () =>
  jest.fn(({ children }) => <div data-testid="layout">{children}</div>)
);

jest.mock("react-bootstrap", () => ({
  Row: jest.fn(({ children }) => <div data-testid="row">{children}</div>),
  Col: jest.fn(({ children }) => <div data-testid="col">{children}</div>),
  Container: jest.fn(({ children }) => (
    <div data-tesid="container">{children}</div>
  )),
  Button: jest.fn(({ children }) => (
    <div data-testid="transferAmountButton">{children}</div>
  )),
  Form: jest.fn(({ children }) => <div data-testid="form">{children}</div>),
}));

const localStorageData = {
  email: "annu",
  password: "1234",
};

const localStorageMock = (function () {
  return {
    getItem: function (key) {
      return localStorageData[key];
    },
    removeItem: function () {
      return null;
    },
    setItem: jest.fn(),
  };
})();

describe("AmountTransfer", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  Object.defineProperty(window, "localStorage", { value: localStorageMock });

  it("should render amount and email input element", () => {
    render(
      <Provider store={store}>
        <Router>
          <AmountTransfer />
        </Router>
      </Provider>
    );

    const inputElement1 = screen.getByPlaceholderText(/Email/i);
    expect(inputElement1).toBeInTheDocument();
    const inputElement2 = screen.getByPlaceholderText(/1000/i);
    expect(inputElement2).toBeInTheDocument();
  });

  it("should render email and amount label element", () => {
    render(
      <Provider store={store}>
        <Router>
          <AmountTransfer />
        </Router>
      </Provider>
    );

    const labeElement1 = screen.getByText("Email");
    expect(labeElement1).toBeInTheDocument();
    const labeElement2 = screen.getByText("Amount");
    expect(labeElement2).toBeInTheDocument();
  });

  it("should render Amount Transfer button", () => {
    render(
      <Provider store={store}>
        <Router>
          <AmountTransfer />
        </Router>
      </Provider>
    );

    const buttonElement = screen.getByTestId("transferAmountButton");
    expect(buttonElement).toBeInTheDocument();
  });

  it("renders properly", () => {
    const { container } = render(
      <Provider store={store}>
        <Router>
          <AmountTransfer />
        </Router>
      </Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should be able to type amount transfer details", () => {
    render(
      <Provider store={store}>
        <Router>
          <AmountTransfer />
        </Router>
      </Provider>
    );

    const textAreaElement1 = screen.getByPlaceholderText(/Email/i);
    fireEvent.change(textAreaElement1, { target: { value: "annu@gmail.com" } });
    expect(textAreaElement1.value).toBe("annu@gmail.com");

    const textAreaElement2 = screen.getByPlaceholderText(/1000/i);
    fireEvent.change(textAreaElement2, { target: { value: "1000" } });
    expect(textAreaElement2.value).toBe("1000");
  });

  describe("#getItem", () => {
    it("returns the user email", () => {
      const result = localStorage.getItem("email");
      expect(result).toEqual("annu");
    });
  });

  describe("when user puts insufficient amount", () => {
    it("Shows insufficient amount", () => {
      const { container, queryByTestId } = render(
        <Provider store={store}>
          <Router>
            <AmountTransfer />
          </Router>
        </Provider>
      );
      act(() => {
        const emailElement = screen.queryByTestId("emailInputBox");
        const amountElement = screen.queryByTestId("amountInputBox");
        fireEvent.change(emailElement, { target: { value: "annu@gmail.com" } });
        fireEvent.change(amountElement, { target: { value: 20000 } });
        const { onSubmit } = Form.mock.calls[2][0];
        onSubmit({ preventDefault: jest.fn() });
      });
      expect(queryByTestId("insufficient-amount")).toBeVisible();
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe("when user puts insufficient amount", () => {
    it("Shows insufficient amount", () => {
      const { queryByTestId } = render(
        <Provider store={store}>
          <Router>
            <AmountTransfer />
          </Router>
        </Provider>
      );
      act(() => {
        const emailElement = screen.queryByTestId("emailInputBox");
        const amountElement = screen.queryByTestId("amountInputBox");
        fireEvent.change(emailElement, { target: { value: "annu@gmail.com" } });
        fireEvent.change(amountElement, { target: { value: 20000 } });
        const { onSubmit } = Form.mock.calls[2][0];
        onSubmit({ preventDefault: jest.fn() });
      });
      expect(queryByTestId("insufficient-amount")).toBeVisible();
    });
  });

  describe("when user puts 0 amount", () => {
    it("Shows error message", () => {
      const { queryByTestId } = render(
        <Provider store={store}>
          <Router>
            <AmountTransfer />
          </Router>
        </Provider>
      );
      act(() => {
        const emailElement = screen.queryByTestId("emailInputBox");
        const amountElement = screen.queryByTestId("amountInputBox");
        fireEvent.change(emailElement, { target: { value: "annu@gmail.com" } });
        fireEvent.change(amountElement, { target: { value: 0 } });
        const { onSubmit } = Form.mock.calls[2][0];
        onSubmit({ preventDefault: jest.fn() });
      });
      expect(queryByTestId("error-msg")).toBeVisible();
    });
  });

  describe("when user puts valid amount", () => {
    it("dispatches proper action", () => {
      const { queryByTestId } = render(
        <Provider store={store}>
          <Router>
            <AmountTransfer />
          </Router>
        </Provider>
      );
      act(() => {
        const emailElement = screen.queryByTestId("emailInputBox");
        const amountElement = screen.queryByTestId("amountInputBox");
        fireEvent.change(emailElement, { target: { value: "annu@gmail.com" } });
        fireEvent.change(amountElement, { target: { value: 500 } });
        const { onSubmit } = Form.mock.calls[2][0];
        onSubmit({ preventDefault: jest.fn() });
      });
      expect(mockDispatch).toBeCalledWith(amountTransfer());
      expect(amountTransfer).toBeCalledWith({
        email: "annu",
        amount: "500",
        creditToEmail: "annu@gmail.com",
        currentAmount: 1000,
      });
    });
  });
});
