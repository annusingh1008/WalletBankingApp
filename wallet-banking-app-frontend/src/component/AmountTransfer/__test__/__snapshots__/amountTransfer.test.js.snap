// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmountTransfer renders properly 1`] = `
<div
  data-testid="layout"
>
  <div
    data-tesid="container"
  >
    <div
      data-testid="row"
    >
      <div
        data-testid="col"
      >
        <div
          class="heading"
        >
          Transfer Amount
        </div>
        <div
          data-testid="form"
        >
          <label>
            Email
          </label>
          <br />
          <input
            data-testid="emailInputBox"
            placeholder="Email"
            required=""
            type="email"
            value=""
          />
          <br />
          <label>
            Amount
          </label>
          <br />
          <input
            data-testid="amountInputBox"
            placeholder="1000"
            required=""
            type="text"
            value=""
          />
          <br />
          <div
            data-testid="transferAmountButton"
          >
            Transfer Amount
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AmountTransfer when user puts insufficient amount Shows insufficient amount 1`] = `
<div
  data-testid="layout"
>
  <div
    data-tesid="container"
  >
    <div
      data-testid="row"
    >
      <div
        data-testid="col"
      >
        <div
          class="heading"
        >
          Transfer Amount
        </div>
        <div
          data-testid="form"
        >
          <label>
            Email
          </label>
          <br />
          <input
            data-testid="emailInputBox"
            placeholder="Email"
            required=""
            type="email"
            value="annu@gmail.com"
          />
          <br />
          <label>
            Amount
          </label>
          <br />
          <input
            data-testid="amountInputBox"
            placeholder="1000"
            required=""
            type="text"
            value="20000"
          />
          <br />
          <p
            class="text-danger"
            data-testid="insufficient-amount"
          >
            Insufficient Amount
          </p>
          <div
            data-testid="transferAmountButton"
          >
            Transfer Amount
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
